birthday.csv파일

#3. 생일 데이터를 교재 코드(heap.py)를 이용해 힙으로 저장하고, 생일이 느린 순서로 10명의 친구를 출력하는 코드를 작성한다. 실행 결과가 셀에 나타나야 한다.

In [ ]:
class Heap:
    def __init__(self, *args):
        if len(args) != 0:
            self.__A = args[0]
        else:
            self.__A = []

    def insert(self, x):
        self.__A.append(x)
        self.__percolateUp(len(self.__A) - 1)

    def __percolateUp(self, i: int):
        parent = (i - 1) // 2
        if i > 0 and self.__A[i] > self.__A[parent]:
            self.__A[i], self.__A[parent] = self.__A[parent], self.__A[i]
            self.__percolateUp(parent)

    def deleteMax(self):
        if not self.isEmpty():
            max = self.__A[0]
            self.__A[0] = self.__A.pop()
            self.__percolateDown(0)
            return max
        else:
            return None

    def __percolateDown(self, i: int):
        child = 2 * i + 1
        right = 2 * i + 2
        if child <= len(self.__A) - 1:
            if right <= len(self.__A) - 1 and self.__A[child] < self.__A[right]:
                child = right
            if self.__A[i] < self.__A[child]:
                self.__A[i], self.__A[child] = self.__A[child], self.__A[i]
                self.__percolateDown(child)

    def max(self):
        return self.__A[0]

    def buildHeap(self):
        for i in range((len(self.__A) - 2) // 2, -1, -1):
            self.__percolateDown(i)

    def isEmpty(self) -> bool:
        return len(self.__A) == 0

    def clear(self):
        self.__A = []

    def size(self) -> int:
        return len(self.__A)

import urllib.request
import csv

url = "https://raw.githubusercontent.com/o9ioakd/250416/main/birthday.csv"
file_path = "birthday.csv"

urllib.request.urlretrieve(url, file_path)

heap = Heap()

with open(file_path, encoding='utf-8') as f:
    reader = csv.reader(f)
    next(reader)
    for row in reader:
        name = row[1]
        birth = row[2]
        heap.insert((birth, name))

print("생일이 느린 순으로 10명:")
for _ in range(10):
    if not heap.isEmpty():
        birth, name = heap.deleteMax()
        print(f"{name}, {birth}")


생일이 느린 순으로 10명:
홍서연, 20241282
신수민, 20051230
이서영, 20051225
강민주, 20051214
김민경, 20051202
이서영, 20051112
배시은, 20051102
김여원, 20051031
이서진, 20051028
서홍빈, 20051024


#4. 생일 데이터를 교재 코드(circularDoublyLinkedList.py)를 이용해 리스트로 저장하고, 같은 조(지난 과제 지정 조원 참조)의 친구들만 이름과 생년월일을 출력하는 코드를 작성한다. 실행 결과가 셀에 나타나야 한다.

In [ ]:
class BidirectNode:
    def __init__(self, x, prevNode:'BidirectNode', nextNode:'BidirectNode'):
        self.item = x
        self.prev = prevNode
        self.next = nextNode

class CircularDoublyLinkedList:
    def __init__(self):
        self.__head = BidirectNode("dummy", None, None)
        self.__head.prev = self.__head
        self.__head.next = self.__head
        self.__numItems = 0

    def insert(self, i:int, newItem) -> None:
        if (i >= 0 and i <= self.__numItems):
            prev = self.getNode(i - 1)
            newNode = BidirectNode(newItem, prev, prev.next)
            newNode.next.prev = newNode
            prev.next = newNode
            self.__numItems += 1
        else:
            print("index", i, ": out of bound in insert()")

    def append(self, newItem) -> None:
        prev = self.__head.prev
        newNode = BidirectNode(newItem, prev, self.__head)
        prev.next = newNode
        self.__head.prev = newNode
        self.__numItems += 1

    def pop(self, *args):
        if self.isEmpty():
            return None
        if len(args) != 0:
            i = args[0]
        if len(args) == 0 or i == -1:
            i = self.__numItems - 1
        if (i >= 0 and i <= self.__numItems - 1):
            curr = self.getNode(i)
            retItem = curr.item
            curr.prev.next = curr.next
            curr.next.prev = curr.prev
            self.__numItems -= 1
            return retItem
        else:
            return None

    def remove(self, x):
        curr = self.__findNode(x)
        if curr != None:
            curr.prev.next = curr.next
            curr.next.prev = curr.prev
            self.__numItems -= 1
            return x
        else:
            return None

    def get(self, *args):
        if self.isEmpty():
            return None
        if len(args) != 0:
            i = args[0]
        if len(args) == 0 or i == -1:
            i = self.__numItems - 1
        if (i >= 0 and i <= self.__numItems - 1):
            return self.getNode(i).item
        else:
            return None

    def index(self, x) -> int:
        cnt = 0
        for element in self:
            if element == x:
                return cnt
            cnt += 1
        return -12345

    def isEmpty(self) -> bool:
        return self.__numItems == 0

    def size(self) -> int:
        return self.__numItems

    def clear(self):
        self.__head = BidirectNode("dummy", None, None)
        self.__head.prev = self.__head
        self.__head.next = self.__head
        self.__numItems = 0

    def count(self, x) -> int:
        cnt = 0
        for element in self:
            if element == x:
                cnt += 1
        return cnt

    def extend(self, a):
        for element in a:
            self.append(element)

    def copy(self) -> 'CircularDoublyLinkedList':
        a = CircularDoublyLinkedList()
        for element in self:
            a.append(element)
        return a

    def reverse(self) -> None:
        prev = self.__head; curr = prev.next; next = curr.next
        self.__head.next = prev.prev; self.__head.prev = curr
        for i in range(self.__numItems):
            curr.next = prev; curr.prev = next
            prev = curr; curr = next; next = next.next

    def sort(self) -> None:
        a = []
        for element in self:
            a.append(element)
        a.sort()
        self.clear()
        for element in a:
            self.append(element)

    def __findNode(self, x) -> BidirectNode:
        curr = self.__head.next
        while curr != self.__head:
            if curr.item == x:
                return curr
            else:
                curr = curr.next
        return None

    def getNode(self, i:int) -> BidirectNode:
        curr = self.__head
        for index in range(i + 1):
            curr = curr.next
        return curr

    def printList(self) -> None:
        for element in self:
            print(element, end=' ')
        print()

    def __iter__(self):
        return CircularDoublyLinkedListIterator(self)


class CircularDoublyLinkedListIterator:
    def __init__(self, alist):
        self.__head = alist.getNode(-1)
        self.iterPosition = self.__head.next


    def __next__(self):
        if self.iterPosition == self.__head:
            raise StopIteration
        else:
            item = self.iterPosition.item
            self.iterPosition = self.iterPosition.next
            return item

    def __iter__(self):
        return self


import csv
import urllib.request

url = "https://raw.githubusercontent.com/o9ioakd/250416/main/birthday.csv"
file_path = "birthday.csv"

urllib.request.urlretrieve(url, file_path)

my_team = [
    "송민서", "안수민", "오예준", "김채민", "김예빈", "신희영", 
    "김선민", "김혜인", "김주하", "김민주", "최가온", "배시은", 
    "강수아", "김서빈"
]

cdll = CircularDoublyLinkedList()

with open(file_path, encoding='utf-8') as f:
    reader = csv.DictReader(f)
    for row in reader:
        name = row['이름'].strip()
        birth = row['생년월일8자리(예.20040101)'].strip()
        if name and birth:
            cdll.append((name, birth))

print("같은 조 친구들 (생일 데이터 있는 경우만):")
for name, birthday in cdll:
    if name in my_team and len(birthday) == 8 and birthday.isdigit():
        print(name, birthday)

같은 조 친구들 (생일 데이터 있는 경우만):
강수아 20041103
김민주 20040517
김민주 20041026
김서빈 20040628
김예빈 20051019
김주하 20050417
김채민 20050910
김혜인 20051001
배시은 20051102
송민서 20041108
신희영 20050126
안수민 20030603
오예준 20050712
최가온 20051008

#5. 교재 8장 우선 순위 큐 연습문제를 모두 푼다.

##1
있을 수 없다.
최대 힙에서는 부모 노드 ≥ 자식 노드의 특성이 유지되어야 하므로,
더 얕은 위치에 있는 노드는 항상 더 크거나 같아야 한다.

##2
항상 그런 것은 아니다.
A[n-1]은 단지 마지막 위치에 있는 원소일 뿐, 가장 작은 값이라는 보장은 없다.

##3
buildHeap은 마지막 부모 노드부터 루트까지만 스며내리기 수행한다.

인덱스 기준: i = (n//2 - 1)부터 0까지
따라서 n개의 노드 중 절반 (n//2개)는 확인하지 않는다.

n - n//2 = ⌊n/2⌋ 개의 노드는 스며내리기 대상이 아니다.

##4
스며내리기는 최악의 경우 힙의 높이만큼 내려가야 한다.

힙의 높이 = log n (완전이진트리 기준)
최악의 경우 (최대 깊이까지 이동): Θ(log n)
최선의 경우 (한 번도 이동하지 않음): Θ(1)

##5
간단하다.
힙의 마지막 원소는 배열의 맨 뒤에 있으므로, pop()을 하면 된다.

##6
더 비효율적이다.
위에서부터 스며오르기를 한다면 각 노드마다 최대 log n 번의 비교를 하게 되며,
n개 노드 전부에 대해 그렇게 하면 → Θ(n log n) 이 됨.

하지만 기존 방법은 Θ(n) 시간에 힙을 만들 수 있다.
따라서 스며내리기 방식이 더 효율적이다.

##7
값이 증가했으므로, 부모보다 커질 수 있다 → 힙 성질 위반 가능
따라서, 아래 과정에 따라 해당 노드를 위로 올려야 한다.


해당 노드의 값을 새 값으로 변경
부모와 비교하면서, 부모보다 크면 swap
root까지 반복 (또는 조건 만족할 때까지)
최대 높이 만큼 → O(log n)

#6. LeetCode 703.Kth Largest Element in Stream 문제를 푼다.

In [ ]:
class MinHeap:
    def __init__(self):
        self.heap = []

    def push(self, val):
        self.heap.append(val)
        self._heapify_up(len(self.heap) - 1)

    def pop(self):
        if len(self.heap) == 0:
            return None
        if len(self.heap) == 1:
            return self.heap.pop()

        top = self.heap[0]
        # Swap first and last, then remove last
        self.heap[0] = self.heap.pop()
        self._heapify_down(0)
        return top

    def peek(self):
        return self.heap[0] if self.heap else None

    def size(self):
        return len(self.heap)

    def _heapify_up(self, index):
        while index > 0:
            parent = (index - 1) // 2
            if self.heap[index] < self.heap[parent]:
                self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
                index = parent
            else:
                break

    def _heapify_down(self, index):
        size = len(self.heap)
        while index < size:
            smallest = index
            left = 2 * index + 1
            right = 2 * index + 2

            if left < size and self.heap[left] < self.heap[smallest]:
                smallest = left
            if right < size and self.heap[right] < self.heap[smallest]:
                smallest = right

            if smallest != index:
                self.heap[index], self.heap[smallest] = self.heap[smallest], self.heap[index]
                index = smallest
            else:
                break

class KthLargest:
    def __init__(self, k: int, nums: list[int]):
        self.k = k
        self.min_heap = MinHeap()
        for num in nums:
            self.add(num)

    def add(self, val: int) -> int:
        self.min_heap.push(val)
        if self.min_heap.size() > self.k:
            self.min_heap.pop()
        return self.min_heap.peek()

#예시 출력
k = 3
nums = [4, 5, 8, 2]
kthLargest = KthLargest(k, nums)

print(kthLargest.add(3))
print(kthLargest.add(5))
print(kthLargest.add(10))
print(kthLargest.add(9))
print(kthLargest.add(4))


4
5
5
8
8